Ignore SIGPIPE events

jwhonce opened the pull request:
Using golang 1.6, is it now possible to ignore SIGPIPE events on stdout/stderr. 
Previous versions of the golang library cached 10 events and then killed the 
process receiving the events.

systemd-journald sends SIGPIPE events when jounald is restarted and
the target of the unit file writes to stdout/stderr. Docker logs to stdout/stderr.

This patch silently ignores all SIGPIPE events.

This resolves #19728

To reproduce issue:
- systemctl start docker
- systemctl restart systemd-journald
- docker run -it --rm kubernetes/pause

Signed-off-by: Jhon Honce jhonce@redhat.com

runcom commented:
Ping @LK4D4 

LGTM

thaJeztah commented:
We're currently not on Go 1.6 yet (we were, but reverted to 1.5.4, due to incompatibilities with older clients, see  https://github.com/docker/docker/issues/20865), so does this work?

runcom commented:
This is basically a no-op on Go < 1.6

cpuguy83 commented:
This is pretty gross.

runcom commented:
@cpuguy83 any other idea how the same could be done (ignoring SIGPIPE and docker daemon kill)

runcom commented:
This has been worked based on the linked issue and comment https://github.com/docker/docker/issues/19728#issuecomment-192044258

googling "ignoring signals in golang" doesn't give us any more option in doing so - if anyone has an idea..

LK4D4 commented:
It is gross indeed, but silent daemon crash is gross too.
Design lgtm

LK4D4 commented on cmd/dockerd/docker.go:
```diff
@@ -23,6 +25,17 @@ func main() {
+	// the docker daemon is not restarted and also running under systemd.
+	go func() {
+		c := make(chan os.Signal, 2048)
+		signal.Notify(c, os.Signal(syscall.SIGPIPE))
```
just `for range c`

LK4D4 commented on cmd/dockerd/docker.go:
```diff
@@ -23,6 +25,17 @@ func main() {
+	// Ignore SIGPIPE events, that are generated by systemd when journald is restarted while
+	// the docker daemon is not restarted and also running under systemd.
+	go func() {
+		c := make(chan os.Signal, 2048)
```
No need in buffered channel imo.

LK4D4 commented:
Also I think it's better to write short commit message title and move whole text to commit message body.

runcom commented:
@jwhonce 
```
19:04:58 ---> Making bundle: validate-gofmt (in bundles/1.12.0-dev/validate-gofmt)
19:04:58 These files are not properly gofmt'd:
19:04:58  - cmd/dockerd/docker.go
19:04:58 
19:04:58 Please reformat the above files using "gofmt -s -w" and commit the result.
```

justincormack commented:
SIGPIPE was a terrible idea. NetBSD has the sane `O_NOSIGPIPE` flag to `open`/`fcntl` which I think the other BSDs are adopting, but Linux does not annoyingly. This seems like the best workaround.

LK4D4 commented on cmd/dockerd/docker.go:
```diff
@@ -23,6 +25,16 @@ func main() {
+	go func() {
+		c := make(chan os.Signal)
+		signal.Notify(c, os.Signal(syscall.SIGPIPE))
+		for range c {
```
I still think that we should log them, on WARN level maybe.

jwhonce replied:
@LK4D4 I didn't log any message to handle the use case where the logging is going to stdout/stderr.  Logging here in that case would create an infinite loop.

jwhonce replied:
Just to be clear if you log a message in this handler:

Docker catches the SIGPIPE, would log to stdout/stderr, systemd would generate a SIGPIPE event, and the whole cycle repeats.

crosbymichael commented:
You move this here:
https://github.com/docker/docker/blob/master/pkg/signal/trap.go#L25

we don't want to have multiple signal handlers scattered throughout the code base. Its better to keep them all in one place.

jwhonce commented:
@crosbymichael I can move this code to signal.Trap() that would also require the call to signal.Trap() to be called earlier in daemon.start().  Otherwise, one of the logrus.Warn/Info calls currently in start() would cause a SIGPIPE event to be generated.

jwhonce committed:
Ignore SIGPIPE events, resolves #19728

Using golang 1.6, is it now possible to ignore SIGPIPE events on
stdout/stderr.  Previous versions of the golang library cached 10
events and then killed the process receiving the events.

systemd-journald sends SIGPIPE events when jounald is restarted and
the target of the unit file writes to stdout/stderr. Docker logs to stdout/stderr.

This patch silently ignores all SIGPIPE events.

Signed-off-by: Jhon Honce <jhonce@redhat.com>

icecrime commented:
@crosbymichael @LK4D4 PTAL!

LK4D4 commented on pkg/signal/trap.go:
```diff
@@ -18,38 +18,43 @@ import (
-							// Call the provided cleanup handler
-							cleanup()
-							os.Exit(0)
+				if sig != syscall.SIGPIPE {
```
check it outside goroutine:
```
if sig == syscall.SIGPIPE {
    continue
}
```

jwhonce commented:
@LK4D4 I implemented your suggestion

LK4D4 commented on pkg/signal/trap.go:
```diff
@@ -18,15 +18,22 @@ import (
 	go func() {
 		interruptCount := uint32(0)
 		for sig := range c {
+			if sig != syscall.SIGPIPE {
```
I think it should be `sig == syscall.SIGPIPE`.

thaJeztah commented:
ping @crosbymichael @LK4D4 PTAL, looks like your comments were addressed; is this ready to merge?

LK4D4 commented:
LGTM

runcom commented:
LGTM

thaJeztah commented:
Looks like we have two LGTM's and all green, so merging
